### PR TITLE
Add ability to clip `Stepper` step content

### LINKED ISSUE #66007: Dismissible content overlays Stepper interface while dismissing it
I'm reporting unexpected effect while swiping Dismissible tile inside a Stepper widget. The Dismissible content can be swiped perfectly but overlays its Stepper widget.

## Steps to Reproduce

1. Below is the code with Stepper (the unexpected effect of using Dismissible inside Stepper).

<details><summary>code</summary>

```dart
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

// MyApp is a StatefulWidget. This allows updating the state of the
// widget when an item is removed.
class MyApp extends StatefulWidget {
  MyApp({Key key}) : super(key: key);

  @override
  MyAppState createState() {
    return MyAppState();
  }
}

class MyAppState extends State<MyApp> {
  final items = List<String>.generate(20, (i) => "Item ${i + 1}");

  @override
  Widget build(BuildContext context) {
    final title = 'Dismissing Items';

    return MaterialApp(
      title: title,
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Scaffold(
        appBar: AppBar(
          title: Text(title),
        ),
        body: Theme(
        data: Theme.of(context).copyWith(
          primaryColor: Colors.red,
        ),
        child: Stepper(
          currentStep: 0,
          steps: [
            Step(
              isActive: true,
              title: Text('Step 1'),
              content: Container(
                padding: const EdgeInsets.all(20),
                color: Colors.black12,
                child: ListView.builder(
                  itemCount: items.length,
                  shrinkWrap: true,
                  itemBuilder: (context, index) {
                    final item = items[index];
                    return Dismissible(
                      key: Key(item),
                      onDismissed: (direction) {
                        setState(() {
                          items.removeAt(index);
                        });
                        Scaffold.of(context).showSnackBar(
                            SnackBar(content: Text("$item dismissed")));
                      },
                      background: Container(color: Colors.red),
                      child: ListTile(title: Text('$item')),
                    );
                  },
                ),
              ),
            ),
            Step(
              title: Text('Step 2'),
              content: Text('content'),
            ),
          ],
        ),
      ),
      ),
    );
  }
}
```
</details>

2. And below is the example code of Dismissible from the docs with padding addition using Container (the expected effect).

<details><summary>code</summary>

```dart
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

// MyApp is a StatefulWidget. This allows updating the state of the
// widget when an item is removed.
class MyApp extends StatefulWidget {
  MyApp({Key key}) : super(key: key);

  @override
  MyAppState createState() {
    return MyAppState();
  }
}

class MyAppState extends State<MyApp> {
  final items = List<String>.generate(20, (i) => "Item ${i + 1}");

  @override
  Widget build(BuildContext context) {
    final title = 'Dismissing Items';

    return MaterialApp(
      title: title,
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Scaffold(
        appBar: AppBar(
          title: Text(title),
        ),
        body: Container(padding: EdgeInsets.all(20), color: Colors.black12, child: ListView.builder(
          itemCount: items.length,
          itemBuilder: (context, index) {
            final item = items[index];

            return Dismissible(
              // Each Dismissible must contain a Key. Keys allow Flutter to
              // uniquely identify widgets.
              key: Key(item),
              // Provide a function that tells the app
              // what to do after an item has been swiped away.
              onDismissed: (direction) {
                // Remove the item from the data source.
                setState(() {
                  items.removeAt(index);
                });

                // Then show a snackbar.
                Scaffold.of(context)
                    .showSnackBar(SnackBar(content: Text("$item dismissed")));
              },
              // Show a red background as the item is swiped away.
              background: Container(color: Colors.red),
              child: ListTile(title: Container(color: Colors.black12, child: Text('$item')),)
            );
          },
        ),
      ),),
    );
  }
}
```
</details>

**Expected results:** 
Same behavior as without Stepper, the content should be swiped without overlaying the parent widget like Container or Padding.
Screenshot (16)

**Actual results:** 
The content overlays the parent widget, which is Stepper and doesn't do the same effect as expected.
Screenshot (13)

<details>

```
[√] Flutter (Channel stable, 1.20.4, on Microsoft Windows [Version 10.0.18363.1082], locale en-ID)
    • Flutter version 1.20.4 at C:\flutter
    • Framework revision fba99f6cf9 (2 days ago), 2020-09-14 15:32:52 -0700
    • Engine revision d1bc06f032
    • Dart version 2.9.2

 
[√] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at C:\Android\Sdk
    • Platform android-29, build-tools 29.0.3
    • ANDROID_HOME = C:\Android\Sdk
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)
    • All Android licenses accepted.

[√] Android Studio (version 4.0)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin version 46.0.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)

[√] VS Code, 64-bit edition (version 1.49.0)
    • VS Code at C:\Program Files\Microsoft VS Code
    • Flutter extension version 3.14.1

[√] Connected device (1 available)
    • Redmi Note 8 (mobile) • 152aa503 • android-arm64 • Android 9 (API 28)

• No issues found!
```

</details>

### PR DESCRIPTION
**Pull request by TahaTesser**
fixes [Dismissible content overlays Stepper interface while dismissing it](https://github.com/flutter/flutter/issues/66007)

### Code sample

<details>
<summary>expand to view the code sample</summary> 

```dart
import 'package:flutter/material.dart';

void main() => runApp(const MyApp());

class MyApp extends StatefulWidget {
  const MyApp({super.key});

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  final List<String> items =
      List<String>.generate(20, (int i) => 'Item ${i + 1}');

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: Padding(
          padding: const EdgeInsets.all(20),
          child: DecoratedBox(
            decoration: BoxDecoration(
              border: Border.all(color: Colors.amber, width: 2),
            ),
            child: Padding(
              padding: const EdgeInsets.all(2.0),
              child: Column(
                children: <Widget>[
                  const SizedBox(height: 8.0),
                  Text(
                    'Dismissible Widget - Vertical Stepper Widget',
                    style: Theme.of(context).textTheme.titleLarge,
                  ),
                  Expanded(
                    child: Stepper(
                      clipBehavior: Clip.hardEdge,
                      steps: <Step>[
                        Step(
                          isActive: true,
                          title: const Text('Step 1'),
                          content: ColoredBox(
                            color: Colors.black12,
                            child: ListView.builder(
                              itemCount: items.length,
                              shrinkWrap: true,
                              itemBuilder: (BuildContext context, int index) {
                                final String item = items[index];
                                return Dismissible(
                                  key: Key(item),
                                  onDismissed: (DismissDirection direction) {
                                    setState(() {
                                      items.removeAt(index);
                                    });
                                    ScaffoldMessenger.of(context).showSnackBar(
                                        SnackBar(
                                            content: Text('$item dismissed')));
                                  },
                                  background: Container(color: Colors.red),
                                  child: ListTile(title: Text(item)),
                                );
                              },
                            ),
                          ),
                        ),
                        const Step(
                          title: Text('Step 2'),
                          content: Text('content'),
                        ),
                      ],
                    ),
                  ),
                  const Divider(height: 1),
                  const SizedBox(height: 8.0),
                  Text(
                    'Dismissible Widget - Horizontal Stepper Widget',
                    style: Theme.of(context).textTheme.titleLarge,
                  ),
                  Expanded(
                    child: Stepper(
                      clipBehavior: Clip.hardEdge,
                      type: StepperType.horizontal,
                      elevation: 0.0,
                      steps: <Step>[
                        Step(
                          isActive: true,
                          title: const Text('Step 1'),
                          content: ColoredBox(
                            color: Colors.black12,
                            child: ListView.builder(
                              itemCount: items.length,
                              shrinkWrap: true,
                              itemBuilder: (BuildContext context, int index) {
                                final String item = items[index];
                                return Dismissible(
                                  key: Key(item),
                                  onDismissed: (DismissDirection direction) {
                                    setState(() {
                                      items.removeAt(index);
                                    });
                                    ScaffoldMessenger.of(context).showSnackBar(
                                        SnackBar(
                                            content: Text('$item dismissed')));
                                  },
                                  background: Container(color: Colors.red),
                                  child: ListTile(title: Text(item)),
                                );
                              },
                            ),
                          ),
                        ),
                        const Step(
                          title: Text('Step 2'),
                          content: Text('content'),
                        ),
                      ],
                    ),
                  ),
                ],
              ),
            ),
          ),
        ),
      ),
    );
  }
}
```

</details>

### Without `Stepper` step content clipping

Group 1

### With `Stepper` step content clipping

Group 2

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

File: packages/flutter/lib/src/material/stepper.dart [hunk 1]
```diff
@@ -366,6 +367,15 @@ class Stepper extends StatefulWidget {
...
   final EdgeInsets? stepIconMargin;
 
+  /// The [Step.content] will be clipped to this Clip type.
+  ///
+  /// Defaults to [Clip.none].
+  ///
+  /// See also:
+  ///
+  ///  * [Clip], which explains how to use this property.
+  final Clip? clipBehavior;
```
Comment (Piinks) [inline packages/flutter/lib/src/material/stepper.dart:377]: Why is this nullable?

File: packages/flutter/test/material/stepper_test.dart [hunk 2]
```diff
@@ -1720,6 +1720,59 @@ testWidgets('Stepper custom indexed controls test', (WidgetTester tester) async
     ));
     expect(lastConnector.width, equals(0.0));
   });
+
+  // This is a regression test for https://github.com/flutter/flutter/issues/66007.
+  testWidgets('Vertical Stepper steps can be clipped', (WidgetTester tester) async {
```
Comment (Piinks) [inline packages/flutter/test/material/stepper_test.dart:1725]: Should we write a test for horizontal as well?

Comment (TahaTesser) [reply packages/flutter/test/material/stepper_test.dart:1725]: It's actually incorrect tittle. Initially I planned to test vertical and horizontal in two separate but it was too simple to split into two tests. Corrected the title. :)

File: packages/flutter/lib/src/material/stepper.dart (hunk 1 above)
Comment (TahaTesser) [reply packages/flutter/lib/src/material/stepper.dart:377]: Looking at the existing `clipBehavior` in the framework. some are nullable and some aren't. The main reason for this is a nullable `clipBehavior` can be updated from the widget theme.

Screenshot 2024-07-30 at 19 01 40
Screenshot 2024-07-30 at 19 01 46


Rethinking after your question, maybe this one makes sense to be non-nullable as there is no Stepper theme.

Comment (TahaTesser) [reply packages/flutter/lib/src/material/stepper.dart]: Made it non-nullable and updated tests.

Comment (Piinks) [reply packages/flutter/lib/src/material/stepper.dart]: Thanks for the context! Nullable makes total sense to me when we have a theme - TIL! :)

**Review by Piinks — APPROVED** [review]
Flutter_LGTM

**Pull request merged**